File: src/main/java/org/fountainmc/api/world/Chunk.java
package org.fountainmc.api.world;

import org.fountainmc.api.world.block.BlockState;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

public interface Chunk {

    public int getX();

    public int getZ();

    public World getWorld();

    public default BlockState getBlockAt(BlockPosition pos) {
        checkNotNull(pos, "Null position");
        checkArgument(pos.getWorld() == this.getWorld(), "Position's world %s doesn't equal chunk's world %s", pos.getWorld().getName(),
                this.getWorld().getName());
        return getBlockAt(pos.getX(), pos.getY(), pos.getZ());
    }

    public BlockState getBlockAt(int x, int y, int z);

}

File: src/main/kotlin/org/fountainmc/api/world/Chunk.kt
package org.fountainmc.api.world

import org.fountainmc.api.world.block.BlockState

interface Chunk {

    val x: Int

    val z: Int

    val world: World

    fun getBlockAt(pos: BlockPosition): BlockState {
        pos.requireChunk(this)
        return getBlockAt(pos.relativeX, pos.y, pos.relativeZ)
    }

    fun getBlockAt(x: Int, y: Int, z: Int): BlockState

}
